### ISSUE TITLE
create_accept_header should treat an empty string header as missing

### ISSUE THREAD

**Issue by mmerickel** [Pylons/webob#366]
```python
>>> create_accept_header(None).acceptable_offers(['foo/bar'])
[('foo/bar', 1.0)]
>>> create_accept_header('invalid/x/y/z').acceptable_offers(['foo/bar'])
[('foo/bar', 1.0)]
>>> create_accept_header('').acceptable_offers(['foo/bar'])
[]
```

versus webob 1.7.4:

```
>>> MIMEAccept('invalid/x/y/z').best_match(['foo/bar'])
>>> MIMENilAccept().best_match(['foo/bar'])
'foo/bar'
```

The test in https://github.com/Pylons/webob/blob/1.7-branch/webob/acceptparse.py#L349-L353 is counting it as missing if "falsey" and I think that is what should be done in 1.8+.

Note also that an invalid header is a pass-through in the webob 1.8 and accepts nothing in the 1.7.4.

**Comment by mmerickel** [comment]
@whiteroses do you have an opinion on this? I'm pretty on the fence but it is different from the old behavior and so I felt the issue should be raised.

**Comment by mmerickel** [comment]
The analysis here is pretty good IMO: https://stackoverflow.com/questions/12130910/how-to-interpret-empty-http-accept-header

Basically, it's ambiguous in the spec. As far as Pyramid is concerned I think the current approach (empty matches nothing) in Webob is ok because you need to define a fallback view anyway and inside there you can return or raise what you like. If we made the empty match everything then it'd match your other views before the fallback view, changing the order slightly and prioritizing other views you've defined.

**Comment by whiteroses** [comment]
@mmerickel: (Please forgive the length: I thought it'd help if I mapped things out so that we could see clearly what the behaviour should be and what changed from 1.7 to 1.8/between `best_match`/`__contains__` and `acceptable_headers`, to help with this and other issues related to the Accept header.)

From [RFC 7231, 5.3.2.  Accept](https://tools.ietf.org/html/rfc7231#section-5.3.2):

    Accept = #( media-range [ accept-params ] )

The `#` [means 0 or more](https://tools.ietf.org/html/rfc7230#section-7), so an empty string header is a valid header. As Accept and Accept-Encoding are `#` while Accept-Charset and Accept-Language are `1#`, we can only assume that this is intentional. (The Stack Overflow answer refers to RFC2616, which is obsolete &mdash; the relevant sections have changed, and some of the text quoted in the answer has been removed.)

The RFC doesn't offer any guidance at all on what to do with an empty string header. But it does offer this:

> If the header field is
> present in a request and none of the available representations for
> the response have a media type that is listed as acceptable, the
> origin server can either honor the header field by sending a 406 (Not
> Acceptable) response or disregard the header field by treating the
> response as if it is not subject to content negotiation.

Given the lack of other guidance, the safest and most straightforward interpretation is to view both as cases of "none of the available representations for the response have a media type that is listed as acceptable" (because no types are listed as acceptable in the header). Then deciding what `best_match`/`__contains__`/`acceptable_offers` should do hinges on whether we think their responsibility should end before, or include, the "406 or disregard header" part (let's call that the fallback response). I've never seen suggestion that WebOb should handle the 406, so I'll only consider the path of disregarding the header, and come back to the 406 later.

The following table lays out the behaviour in all the possible cases for the header, depending on whether our responsibility ends before the fallback response (in which case we return the offers considered acceptable by the header), or includes the fallback (in which case we disregard the header when no offers are considered acceptable by the header):

|      | Before fallback response | After fallback response (where required) |
| --- | --------------------------------- | ------------------------------- |
| Valid header, one or more offers considered acceptable by header | Return all/first of the offer(s) that are acceptable | Return all/first of the offer(s) that are acceptable |
| Valid header, no offers considered acceptable by header, header is empty string | Return `[]`/`None` | Disregard header, return all/first of the offer(s) |
| Valid header, no offers considered acceptable by header, header is not empty string | Return `[]`/`None` | Disregard header, return all/first of the offer(s) |
| No header in request | Return all/first of the offer(s) | Return all/first of the offer(s) |
| Invalid header | Return `[]`/`None` | Disregard header, return all/first of the offer(s) |

<br>
<br>

The behaviour of `best_match` and `__contains__` in WebOb 1.7.4 is:

|      | Before fallback response | After fallback response (where required) |
| --- |:---------------------------------:|:-------------------------------:|
| Valid header, one or more offers considered acceptable by header | X | X |
| Valid header, no offers considered acceptable by header, header is empty string |  | X |
| Valid header, no offers considered acceptable by header, header is not empty string | X | |
| No header in request | X | X |
| Invalid header\* | X | |

\* MIMEAccept had no concept of an invalid header &mdash; it went through the header looking for valid ranges and ignoring everything else around them, so this case refers to when there are no valid ranges in the header, only invalid content.

    >>> MIMEAccept('a/b, c/d').best_match(['c/d', 'e/f'])
    'c/d'
    >>> MIMENilAccept().best_match(['c/d', 'e/f'])
    'c/d'
    >>> MIMEAccept('a/b').best_match(['c/d', 'e/f']) is None
    True
    >>> MIMEAccept('invalid').best_match(['c/d', 'e/f']) is None
    True

    >>> 'c/d' in MIMEAccept('a/b, c/d')
    True
    >>> 'a/b' in MIMENilAccept()
    True
    >>> 'a/b' in MIMEAccept('c/d, e/f')
    False
    >>> 'a/b' in MIMEAccept('invalid')
    False


<br>
<br>

The behaviour of `acceptable_offers` in WebOb 1.8 is:

|      | Before fallback response | After fallback response (where required) |
| --- |:---------------------------------:|:-------------------------------:|
| Valid header, one or more offers considered acceptable by header | X | X |
| Valid header, no offers considered acceptable by header, header is empty string | X |  |
| Valid header, no offers considered acceptable by header, header is not empty string | X | |
| No header in request | X | X |
| Invalid header | | X |

    >>> create_accept_header('a/b, c/d').acceptable_offers(['c/d', 'e/f'])
    [('c/d', 1.0)]
    >>> create_accept_header('').acceptable_offers(['c/d', 'e/f'])
    []
    >>> create_accept_header('a/b').acceptable_offers(['c/d', 'e/f'])
    []
    >>> create_accept_header(None).acceptable_offers(['c/d', 'e/f'])
    [('c/d', 1.0), ('e/f', 1.0)]
    >>> create_accept_header('invalid').acceptable_offers(['c/d', 'e/f'])
    [('c/d', 1.0), ('e/f', 1.0)]

<br>
<br>

We can see that neither `best_match`/`__contains__` nor `acceptable_offers` follows "before fallback response" *or* "after fallback response" entirely. If `best_match` 

From what I can recall, my intention with `acceptable_offers` was that it would be responsible for what happens before the fallback, and that the rest of the response could be built easily on top of it as required: a `best_match` or a `__contains__`, or a function or method, e.g. a `best_response_offer`, that handles the "disregard header" option of the fallback response:

    def best_response_offer(self, offers):
        acceptable_offers = self.acceptable_offers(offers=offers)
        return acceptable_offers[0][0] if acceptable_offers else offers[0]

I cannot remember why I special-cased the invalid header and made it return the fallback response, and can only assume it was a mistake or a misreading &mdash; I'll submit a PR to correct this. The response for an empty string header is imo the right one.

(`acceptable_offers` stops before the fallback because the user might want to 406 instead of disregarding the header (examples from Pyramid issues of two people who'd like to 406: [1](https://github.com/Pylons/pyramid/issues/1261), [2](https://github.com/Pylons/pyramid/issues/1873)). Or do something else entirely: RFC says "the origin server *can* [emphasis mine] either honor the header field...", not MUST. A method to handle the fallback or return a custom response can be easily built on top of `acceptable_offers`.

`best_match` also seems to be written to stop before fallback: it returns `None` for the common case of "Valid header, no offers considered acceptable by header, header is not empty string", rather than disregard the header and return first of the offers.)

**Comment by mmerickel** [comment]
> If you'd like, I can help adapt acceptable_offers to work with Pyramid

To be clear the only reason I'm doing this work is that no one else has done it and I think it's important to do and was the motivation for offering the Webob portion for GSOC. Of course you're welcome to help with the Pyramid part but that has been the case for some time now and at this point -- at least for the view lookup stage I think I have finished the work other than writing tests and getting these PRs merged into master. As for the http exception stuff, I have not looked at that and you're welcome to open PRs or issues to Pyramid for that.

> The response for an empty string header is imo the right one.

With respect to this issue, I think we all agree now that treating an empty string as a valid header that matches nothing is fine.

> I cannot remember why I special-cased the invalid header and made it return the fallback response, and can only assume it was a mistake or a misreading — I'll submit a PR to correct this.

Feel free to read https://github.com/Pylons/webob/issues/368 and decide that we came to the wrong conclusion and submit a PR. Personally, I don't even see the point of `AcceptInvalidHeader` versus `AcceptValidHeader` and think that the only thing anyone cares about is `AcceptHeader` versus `AcceptMissingHeader`. An invalid header is a header someone passed but is erroneous and thus shouldn't match anything (or should only match the parts we could parse).

>  If we could have a little more time before Pyramid 1.10 is released (or if the changes could wait until the next release), it would give us time to map it out and get it right.

What I really care about right now is challenging the current implementation, getting everyone on the same page, and shipping Pyramid in a way that is documented and expected to continue working in future releases.
